### PR TITLE
feat(2447): implement profile edit (visibility, acc. details, email prefs)

### LINKED ISSUE #2447: Webpage Integration: Edit User Profile - Visibility Toggles, Email Prefs & Update Details
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Description</h2>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">Implement visibility toggles (hide GitHub activity, mailing list activity, badges), account detail fields (username, country, login indicator, commit author), and email notification preferences. Each section has its own "Save changes" button.</p>
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Acceptance Criteria</h2>
<h3 class="text-text-100 mt-2 -mb-1 text-base font-bold">Visibility toggles</h3>
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> Toggle: Hide GitHub activity from public profile</li>
<li class="task-list-item"><input disabled="" type="checkbox"> Toggle: Hide mailing list activity from public profile</li>
<li class="task-list-item"><input disabled="" type="checkbox"> All toggles are saved only when user clicks "Save changes" in Profile section</li>
</ul>
<h3 class="text-text-100 mt-2 -mb-1 text-base font-bold">Account details (Update Details section)</h3>
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> <strong>Username / Display name</strong>: editable text input field
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> Validated by backend on save</li>
<li class="task-list-item"><input disabled="" type="checkbox"> Error message if taken: "This username is already taken"</li>
</ul>
</li>
<li class="task-list-item"><input disabled="" type="checkbox"> <strong>Email</strong>: display-only disabled input (cannot be edited from profile page)</li>
<li class="task-list-item"><input disabled="" type="checkbox"> <strong>Country</strong>: searchable dropdown populated with all countries
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> Selection determines flag shown on public profile and post feed cards</li>
<li class="task-list-item"><input disabled="" type="checkbox"> Selecting blank/no country hides flag display</li>
</ul>
</li>
<li class="task-list-item"><input disabled="" type="checkbox"> <strong>Indicate last login method</strong>: checkbox toggle
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> When checked: login page displays last-used OAuth method (Google, GitHub) for user convenience</li>
</ul>
</li>
<li class="task-list-item"><input disabled="" type="checkbox"> <strong>Override commit author name</strong>: checkbox toggle
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> When checked: globally replaces user's git commit author name with their username</li>
<li class="task-list-item"><input disabled="" type="checkbox"> When unchecked: reverts to original commit author behavior</li>
</ul>
</li>
</ul>
<h3 class="text-text-100 mt-2 -mb-1 text-base font-bold">Email preferences</h3>
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> Users opt in/out via checkboxes for notification types:
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> My content approved: blog posts, links, news, video</li>
<li class="task-list-item"><input disabled="" type="checkbox"> Others publish content: blog posts, links, news, video</li>
</ul>
</li>
<li class="task-list-item"><input disabled="" type="checkbox"> All preferences saved only when user clicks "Save changes" in Email preferences section</li>
</ul>
</li>
note: The site's terms of use or privacy policy are changed (needs to be removed from the UI, no longer part of the page
</ul>
<h3 class="text-text-100 mt-2 -mb-1 text-base font-bold">Backend</h3>
<ul class="contains-task-list">
<li class="task-list-item"><input disabled="" type="checkbox"> Automatically loads existing profile datas</li>
</ul>
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Out of Scope</h2>
<ul class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-disc flex flex-col gap-1 pl-8 mb-3">
<li class="font-claude-response-body whitespace-normal break-words pl-2">Published posts hidability (inherently public once published)</li>
<li class="font-claude-response-body whitespace-normal break-words pl-2">Commit email address management (separate ticket)</li>
<li class="font-claude-response-body whitespace-normal break-words pl-2">Delete account flow (separate ticket)</li>
<li class="font-claude-response-body whitespace-normal break-words pl-2">OAuth account management (separate spike)</li>
</ul>
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Design Reference</h2>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]"><a class="underline underline underline-offset-2 decoration-1 decoration-current/40 hover:decoration-current focus:decoration-current" href="https://www.figma.com/design/5j0fQssrV9ipoU16P7hfKy/Website-Deliverables?node-id=1644-125310">Edit User Profile — Visibility toggles</a><br>
<a class="underline underline underline-offset-2 decoration-1 decoration-current/40 hover:decoration-current focus:decoration-current" href="https://www.figma.com/design/5j0fQssrV9ipoU16P7hfKy/Website-Deliverables?node-id=1644-125310">Edit User Profile — Update account details</a><br>
<a class="underline underline underline-offset-2 decoration-1 decoration-current/40 hover:decoration-current focus:decoration-current" href="https://www.figma.com/design/5j0fQssrV9ipoU16P7hfKy/Website-Deliverables?node-id=1644-125310">Edit User Profile — Email preferences</a></p>
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">

</div></b>

### PR DESCRIPTION
**Pull request by ycanales**
# Issue: #2447

## Summary & Context

Wires the v3 edit-profile page (shell landed in Story 2437) so three of its cards actually validate and persist: profile visibility toggles, account details (username, country), and email preferences. Adds four new `User` fields plus a per-section form pattern so each card saves independently.

- **Figma link**: https://www.figma.com/design/5j0fQssrV9ipoU16P7hfKy/Website-Deliverables?node-id=1644-125601
- **Link to components/page:** localhost:8000/users/me/?edit=true

## Changes

- Added a per-section form pattern: each card is its own `<form>` with a distinct named submit button (`v3_update_profile`, `v3_update_details`, `v3_update_email_preferences`); `post_v3()` dispatches on the button name and validates/saves only that section, so an unfinished card never blocks a completed one. Section wiring lives in the `V3_EDIT_SECTIONS` map in `users/views.py`.
- Added a `dispatch()` override on `CurrentUserProfileView` to enforce login and route v3 POSTs into `post()`/`post_v3()` (the v3 mixin otherwise renders on every request and never reaches `post`), plus `get_v3_edit_initial()`/`get_v3_edit_context()` to seed the form from the current user.
- Added four new fields to the `User` model (migration `0022`): `country` (django-countries `CountryField`), `hide_github_activity`, `hide_mailing_list_activity`, and `hide_badges` (all `BooleanField`).
- Update Details section saves Username (`User.display_name`, with case-insensitive uniqueness in `clean_username`) and Country (`User.country`), plus the existing `indicate_last_login_method` and `is_commit_author_name_overridden` toggles. Country was previously an empty plain dropdown and is now wired to the existing searchable combo (`_field_combo.html`) populated from django-countries.
- Visibility section saves the three new hide toggles (`hide_github` / `hide_ml` / `hide_ach`).
- Email Preferences section saves the two checkbox groups into `Preferences.notifications` (JSONField) via the existing property accessors, using a 4-item choice list (`V3_EMAIL_PREFERENCE_CHOICES`, no Poll to match Figma); unlisted news types a user already had enabled (e.g. Poll) are preserved on save.
- Extended shared includes: `_button.html` accepts a `name=` so multiple submit buttons can coexist on one page, `_field_text.html` / `_field_include.html` surface server-side errors and seed values on re-render, and `card.css` adds `.card__form` so a `<form>` can wrap part of a card without breaking its flex-column layout.
- Added `users/tests/test_v3_profile_edit.py` (11 tests) covering login-required GET/POST, current-value rendering, each section's save, blank-country clearing, duplicate-username inline error, email-prefs save, and the Poll-preservation edge case.

## ‼️ Risks & Considerations ‼️
*Please list any potential risks or areas that need extra attention during review/testing*

- New migration `0022` adds four columns to the `User` table; needs to run on deploy.
- `email`, the Commit email address card, Account connections, and the top of the Profile card (tagline, bio, links, avatar) are still UI-only scaffolding and do not persist yet. The `override_commit_author_email` toggle also has no backing field. These are intentionally out of scope for this PR.
- ~Known visual bug: the Username/Email inputs in Update Details render around 22px tall instead of the 40px in Figma (the `.field__control` height is being overridden somewhere), so they look pill-shaped. Still being diagnosed.~ Fixed.
- Behavior is gated behind the `v3` waffle flag.
- **Username validations:** in this PR, username is only checked for an existing exact duplicate, more extensive rules will be implemented in this follow-up ticket https://github.com/boostorg/website-v2/issues/2521 

## Screenshots

edit_01
localhost_8000_users_me__edit=true

## Self-review Checklist
- [x] Link this PR to the related GitHub Project ticket

### Frontend
- [x] UI implementation matches Figma design
- [x] Tested in light and dark mode
- [x] Responsive / mobile verified
- [x] Accessibility checked (keyboard navigation, etc.)
- [x] Ensure design tokens are used for colors, spacing, typography, etc. – No hardcoded values
- [ ] Test without JavaScript (if applicable): separate ticket will deal with all segments of the page for no-JS functionality.
- [x] No console errors or warnings

## Summary by CodeRabbit

* **New Features**
  * Added v3 section-based profile editing (visibility, account details, email preferences) with per-section saved-state handling and JSON/XHR responses.
  * Added country support plus “hide” controls for GitHub activity, mailing list activity, and badges.
* **Bug Fixes**
  * Improved form validation and error display (duplicate username checks, inline first-error context, and correct max-length behavior for pre-filled values).
  * Preserved existing email preference behavior not managed by the v3 UI.
* **Style**
  * Improved v3 card/form and label-row layout spacing/alignment.
* **Tests**
  * Added v3 profile edit flow coverage behind a feature flag.

### CODE REVIEW

**Comment by ycanales** [comment]
On removing the `height: 100%` on .field__label-row in `static/css/v3/forms.css`:

In the Update Details card, the field sits in an unbroken chain of definite-height containers: the grid stretches .user-profile__col to the row height, and from there card → `.user-profile__xl-space-col` → .field are all column-flex layouts. In that context the label row's height: 100% actually resolves — against the whole .field height — so the label row's flex base size ballooned to ~70px. The field then had more base content than room, and flex-shrink compressed the `.field__control` from its specified 40px down to its ~20px min-content floor (this is why the height looked "ignored": flex shrink overrides even an inline height: 40px !important). 

Tagline and the other inputs never broke because they sit inside plain block divs, which make the parent height indefinite; there height: 100% silently fell back to auto.

Before:
image

After:
image

I double-checked other views to confirm heights were not affected:

- /accounts/login - email field
- /accounts/signup/ - email and username fields
- /v3/news/add/ - title and link fields
- /libraries/ - search field 
- /news/ - search field

File: users/forms.py [hunk 1]
```diff
@@ -240,6 +252,17 @@ def __init__(self, *args, **kwargs):
...
+        username = self.cleaned_data["username"]
+        if not username:
+            return username
+        existing = User.objects.filter(display_name__iexact=username)
+        if self._user is not None:
+            existing = existing.exclude(pk=self._user.pk)
+        if existing.exists():
+            raise forms.ValidationError("This username is already taken")
+        return username
+
```
[the remark this answers is not in the document]
Comment (julhoang) [reply users/forms.py:283]: @ycanales Would you mind confirming if this will be handled in #2521 ?

File: templates/v3/user_profile_edit.html [hunk 2]
```diff
@@ -139,8 +144,10 @@
           </div>
           <hr class="card__hr" aria-hidden="true" />
           <div class="card__cta_section">
-            {% include 'v3/includes/_button.html' with style='primary' type='submit' label='Save Changes' only %}
+            {% include 'v3/includes/_button.html' with style='primary' type='submit' name='v3_update_profile' label='Save changes' only %}
+            {% include 'v3/includes/_button.html' with style='secondary' url=profile_account_url label='Cancel' only %}
```
Comment (julhoang) [inline templates/v3/user_profile_edit.html]: Instead of redirecting user back to "My Profile", I think it'd make a greater UX if we allow users to remain on the Edit User page with just the temporary form data revert. To do that, I think we just need to replace `url=profile_account_url` with `type='reset'` here.

**Review by julhoang — CHANGES_REQUESTED** [review]
Hi @ycanales ! This is great work! I just ran into 1 issue with saving the `Override commit author email` UI field. Would you mind double-checking this?

https://github.com/user-attachments/assets/c4c1898e-4419-40f3-ad00-8a8c13ec04d3

Also, per the [Figma design](https://www.figma.com/design/5j0fQssrV9ipoU16P7hfKy/Website-Deliverables?node-id=6635-27603&t=tIz061L1UPXjVpM6-4), the button text should be updated to `Changes saved` instead of announcing via the toast message. Can we suppress the legacy toast message on this page? 

Image

File: templates/v3/user_profile_edit.html [hunk 3]
```diff
@@ -16,6 +16,11 @@
     <div class="user-profile__profile-content-area">
       <div class="user-profile__col">
         <div class="basic-card card user-profile__profile-edit-card">
+          {% comment %} novalidate: this form still contains required fields (tagline, bio, links) with
+          no save handler yet; native constraint validation would silently block "Save changes" for
+          sections that are otherwise ready. Server-side validation is authoritative. {% endcomment %}
+          <form method="post" action="{% url 'profile-account' %}?edit=true" class="card__form" enctype="multipart/form-data" novalidate>
```
Comment (herzog0) [inline templates/v3/user_profile_edit.html]: Nit: Can we build the edit URL in the backend an hydrate it in the view, please?

File: templates/v3/user_profile_edit.html [hunk 4]
```diff
@@ -161,22 +168,25 @@
...
+          {% csrf_token %}
           <div class="card__header">
             <span class="card__title">Update Details</span>
           </div>
           <hr class="card__hr" aria-hidden="true" />
           <div class="user-profile__xl-space-col px-large">
             {% include 'v3/includes/_field_include.html' with bound_field=user_profile_form.username %}
             {% include 'v3/includes/_field_include.html' with bound_field=user_profile_form.email %}
-            {% include 'v3/includes/_field_include.html' with bound_field=user_profile_form.country %}
+            {% include 'v3/includes/_field_combo.html' with name=user_profile_form.country.name label='Country' options=country_options selected=user_profile_form.country.value placeholder='Select' error=user_profile_form.country.errors.0 %}
```
Comment (herzog0) [inline templates/v3/user_profile_edit.html:203]: Per the ticket, the user should also be able to select a blank/no country option to hide their flag if they don't wanna disclose that info anymore. Currently, there's no way to do that. 

Image

**Review by herzog0 — DISMISSED** [review]
Looking good, just one nit, one requirement and one note: about the "Override commit author email address" feature, don't worry about that (cc @julhoang) as this will be addressed in a dedicated ticket. This feature shouldn't be implemented as this design reference states anyway.

File: templates/v3/user_profile_edit.html (hunk 4 above)
Comment (ycanales) [reply templates/v3/user_profile_edit.html:203]: Good catch! Now when the user selects "No country" the select clears and is persisted when saving the form. Thanks Teo 👍 
image

File: templates/v3/user_profile_edit.html (hunk 3 above)
Comment (ycanales) [reply templates/v3/user_profile_edit.html]: Addressed, thanks!

File: templates/v3/user_profile_edit.html (hunk 2 above)
Comment (ycanales) [reply templates/v3/user_profile_edit.html]: Good catch on the toast/"Changes Saves" button, here it is:
image

If you change again a value, the button changes back:
image


I addressed the "Cancel" behavior too, I wonder though, maybe the Figma is missing the "Cancel" in the other sections too, or could it be intentional?
image

**Comment by ycanales** [comment]
Merged develop into my branch and solved conflicts while keeping changes from both sides.
Migration was changed from 0022 to 0023 and the dependency was updated.

**Comment by ycanales** [comment]
Addressed CodeRabbit suggestion on non-submitted form sections being visually erased on validation failure.

**Review by julhoang — COMMENTED** [review]
Woops we're now having an extra Save button in this section 😆, would you mind double-checking this? 

Image

File: templates/v3/user_profile_edit.html (hunk 4 above)
Comment (julhoang) [reply templates/v3/user_profile_edit.html:203]: We actually do have a `deselectable` version of some dropdowns which were used on the [Libraries page](http://localhost:8000/libraries/1.90.0/list/?grading=flagship&min_cpp=03&max_cpp=17) (I think I only added it to `_field_combo_multi.html`  and `_field_dropdown.html`), I'm curious if we can use the same design here 🤔 That being said, I think the "None" option works great too!
cc @henryajisegiri 

Screenshot 2026-07-16 at 2 20 36 PM

**Comment by ycanales** [comment]
Fixed the extra Save button, it came from the work on the user profile links, now all is handled by the Save button from this PR. Thanks @julhoang !

File: templates/v3/user_profile_edit.html (hunk 4 above)
Comment (ycanales) [reply templates/v3/user_profile_edit.html:203]: Thanks! It is now using the deselectable version, but I also added an option to not change the border color to blue, to keep the current design with grey borders. 
image

File: templates/v3/includes/_field_combo.html
Comment (julhoang) [inline templates/v3/includes/_field_combo.html:1]: After trying the "No Country" option, I think it works really well – so we may not actually need the deselect (X) button and add a small drift from design. My apologies for reversing my earlier suggestion! Is there any way you could help us revert the deselectable variant support here? 🙏

File: templates/v3/user_profile_edit.html [hunk 5]
```diff
@@ -140,17 +158,15 @@
...
-            {% include 'v3/includes/_button.html' with style='primary' type='button' label='Save Changes' alpine_disabled="saving" alpine_click="save()" only %}
-            <span class="user-profile__save-indicator" role="status" aria-live="polite">
-              <span class="user-profile__save-state" x-show="saveStatus === 'saved'" x-cloak>
-                <span class="user-profile__save-text">Saved</span>
-                {% include "includes/icon.html" with icon_name="saved" icon_class="create-post-page__save-icon" icon_size=16 %}
-              </span>
-              <span class="field__error" x-show="saveStatus === 'error'" x-text="saveError" x-cloak></span>
-            </span>
+            {% include 'v3/includes/_button.html' with style='primary' type='submit' name='v3_update_profile' label='Save changes' label_x_text="saved ? 'Changes Saved' : 'Save changes'" alpine_disabled='saved || saving' only %}
+            {% include 'v3/includes/_button.html' with style='secondary' type='reset' label='Cancel' only %}
```
Comment (julhoang) [inline templates/v3/user_profile_edit.html]: After checking with Henry – since we have different design variants for this page – we're going with [this latest version](https://www.figma.com/design/5j0fQssrV9ipoU16P7hfKy/Website-Deliverables?node-id=6635-24737&t=q4O0YVIECMSpYG86-4) that doesn't include Cancel buttons. Again, sorry for the confusion from the design drifts. Would you mind helping us remove this Cancel button here? 🙏

**Review by julhoang — CHANGES_REQUESTED** [review]
Hi @ycanales , I have a 2 design requests below – thanks in advance for your help with correcting the UI 🙏.

Asides from that, I noticed one small issue the Country selector – changes made to this dropdown selection don't get detected as a valid change for the "Save" button to become active. You can follow these steps to reproduce:
1. Refresh the page, and make some changes on the form (e.g. tick a checkbox)
2. Hit "Save" -> This button will now says "Changes Saved" and become disabled
3. Select a country from the Country dropdown 
4. Notice how the button doesn't revert back to "Save details" and become active 
5. Try to change 1 of the checkbox -> Now the button correctly becomes active.

File: templates/v3/includes/_field_combo.html
Comment (ycanales) [reply templates/v3/includes/_field_combo.html:1]: No problem! Thanks for verifying :). I'll revert the `deselectable` change.

File: templates/v3/user_profile_edit.html (hunk 5 above)
Comment (ycanales) [reply templates/v3/user_profile_edit.html]: No problem, thank you for validating with Henry! I'll remove it.

File: templates/v3/includes/_field_combo.html
Comment (ycanales) [reply templates/v3/includes/_field_combo.html:1]: Changed back without `deselectable` and fixed the _Save_ not becoming active the second time you change the country after saving.
Thanks again @julhoang for your feedback!

File: templates/v3/user_profile_edit.html (hunk 5 above)
Comment (ycanales) [reply templates/v3/user_profile_edit.html]: Removed and cleaned up, thanks again @julhoang !

**Review by jlchilders11 — APPROVED** [review]
This code all seems solid to me! I am approving, but am not looking forward to the merge conflicts as we all start to merge these various sections.

**Review by julhoang — COMMENTED** [review]
Heya @ycanales , thank you for the updates you've made! We're super close, it seems like just the Save issue on the Country selector hasn't worked yet. Would you mind resolving that as well?

**Review by julioest — APPROVED** [review]
Approving! Pulled it down and tested the edit-profile flow locally: all three sections (visibility, account details, email preferences) save and persist independently, the per-section state holds up, and the duplicate-username guard works as expected. Looks good.

**Comment by ycanales** [comment]
@julhoang my bad I commented before pushing the changes, now they're up.

**Review by julhoang — APPROVED** [review]
Everything looks great to me, thanks a ton for all the updates @ycanales ! 🙌

**Review by herzog0 — APPROVED** [review]
Working perfectly on my end. Thanks a ton for the fixes @ycanales

**Comment by ycanales** [comment]
Merged latest changes from develop (no change in scope nor behavior @kattyode )

**Review by julhoang — COMMENTED** [review]
Hi @ycanales, I've just happened to notice that right now `users` has two `0022_*` migrations both branching off `0021`, plus a `0024_merge` to reconcile them, which was probably introduced by the merge. 

Could we collapse it back into a linear chain – rename it to `0024_user_country_...`
depending on `0023_user_biography_user_tagline` and drop the empty `0024_merge`? I think it should be a fairly clean one!

**Comment by ycanales** [comment]
@julhoang good call! I had an issue with the migrations deploying to dev but I got it sorted out and should be clean onwards with this now refactored to 0024_user_country...
Thanks!

**Review by kattyode — APPROVED** [review]
Approved
